### PR TITLE
Failure running workflow with inline activity

### LINKED ISSUE #592: Failure running workflow with inline activity
# Bug Report

## Description

Attempting to run a workflow with an inline activity, like workflow wf2 below :
```yaml
    wf1:
      steps:
        Component_say_hello:
          target: Component
          activities:
            - call_operation: custom.say_hello
    wf2:
      steps:
        inline_wf1:
          activities:
            - inline: wf1
```

The execution of workflow wf2 is described to be successful, but  actually the inline workflow wf1 was not executed.
Yorc logs show this error :
```
[ERROR][yorc-example-Environment][wf2][78cfcf58-c218-49fb-94b3-fd173b140d5a][][][][][][]Failed to build step: "Component_say_hello" for workflow: "wf2", unknown step
[DEBUG][yorc-example-Environment][wf2][78cfcf58-c218-49fb-94b3-fd173b140d5a][][][][][][]Failed to build step: "Component_say_hello" for workflow: "wf2", unknown step
github.com/ystia/yorc/v4/tasks/workflow.(*worker).runWorkflowStep
        /home/travis/gopath/src/github.com/ystia/yorc/tasks/workflow/worker.go:777
github.com/ystia/yorc/v4/tasks/workflow.(*worker).runCustomWorkflow
        /home/travis/gopath/src/github.com/ystia/yorc/tasks/workflow/worker.go:748
github.com/ystia/yorc/v4/tasks/workflow.(*worker).handleExecution
        /home/travis/gopath/src/github.com/ystia/yorc/tasks/workflow/worker.go:220
github.com/ystia/yorc/v4/tasks/workflow.(*worker).Start.func1
        /home/travis/gopath/src/github.com/ystia/yorc/tasks/workflow/worker.go:83
runtime.goexit
```

## Output of `yorc version`

4.0.0-SNAPSHOT

## Priority

Medium

* High = We should stop anything we are doing and do this. If you're writing this calmly, you should probably choose another option. If it's really that urgent, please make sure we read this.
* Medium = It's important that we do this within a few days.
* Low = We will consider this on our next sprint planning.

(Please be aware that your priority may not match ours, we'll use this as guidance only).

### PR DESCRIPTION
**Pull request by laurentganne**
# Pull Request description

### What I did

Executing an inline workflow step in a new task, as this is really a new workflow made of steps itself, and running it in a separate task avoids potential step names collisions between the parent workflow and the child workflow that could occur if both were run in the same task.

### How I did it

Changes done:

`tasks/workflow/step.go`
Function registerInlineWorkflow():
Preparing a new task with its own data referencing the parent workflow step and parent task.
This task will take care of executing the inline workflow.

`tasks/workflow/worker.go`
Function runCustomWorkflow():
Modified the final function called when the workflow has finished, to do additional actions if this workflow is an inline workflow launched by a parent workflow step:
* update the parent workflow step status according to the status of this inline workflow execution
* register the parent workflow next steps

`tasks/workflow/workflow.go`:
Added functions to get the name of parent workflow from task data, update a parent workflow step state and register parent workflow next steps

### How to verify it

Using yorc as A4C UI doesn't work very well to follow the execution of a workflow with inline steps.
Deploy this zip file in attachment:
[deployment.zip](https://github.com/ystia/yorc/files/4171059/deployment.zip)
describing an OpenStack application with these custom workflows wf1, and wf2 for which the first step is the inline workflow wf1:
```yaml
    wf1:
      steps:
        Component_say_hello:
          target: Component
          activities:
            - call_operation: custom.say_hello
          on_success:
            - Component_say_goodbye
        Component_say_goodbye:
          target: Component
          activities:
            - call_operation: custom.say_goodbye
    wf2:
      steps:
        inline_wf1:
          activities:
            - inline: wf1
          on_success:
            - Component_say_hello
        Component_say_hello:
          target: Component
          activities:
            - call_operation: custom.say_hello
```

Run this command to deploy the application:

```bash
yorc deployments deploy deployment.zip --id test --stream-logs
```

Once done, run the workflow and follow the events, using this command:

```bash
yorc deployments workflows execute test --workflow-name wf2 --stream-events
```

Check these events occur showing workflow wf2 is running, this triggers the execution of workflow wf1 steps custom.say_hello and custom.say_goobye,
then the execution of workflow wf2 last step custom.say_hello :

```
Deployment: test Task "6d1bd3b6-9bd0-4c4f-8ac4-bf91eb87ea3e" (workflow) Workflow: wf2	 Status: running
[...] (workflow)	 Workflow: wf1	 Status: initial
[...] (workflow)	 Workflow: wf1	 Status: running
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: initial
[...] (Execution)	 Execution: "5605e565-bbbe-4b77-a48b-b96d84f98281-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: initial
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: running
[...] (Execution)	 Execution: "5605e565-bbbe-4b77-a48b-b96d84f98281-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: running
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: done
[...] (Execution)	 Execution: "5605e565-bbbe-4b77-a48b-b96d84f98281-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: done
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: initial
[...] (Execution)	 Execution: "57005bf3-b671-4e90-9fe5-8cdc6edc3ca0-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: initial
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: running
[...] (Execution)	 Execution: "57005bf3-b671-4e90-9fe5-8cdc6edc3ca0-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: running
[...] (workflowStep)	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: done
[...] (Execution)	 Execution: "57005bf3-b671-4e90-9fe5-8cdc6edc3ca0-0"	 Workflow: wf1	 Instance: 0	 Step: Component_say_goodbye	 Node: Component	 Operation: custom.say_goodbye	 Status: done
[...] (workflow)	 Workflow: wf1	 Status: done
[...] (workflowStep)	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: initial
[...] (Execution)	 Execution: "0d00bc24-fb5b-4399-acf2-340c8e3e6c05-0"	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: initial
[...] (workflowStep)	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: running
[...] (Execution)	 Execution: "0d00bc24-fb5b-4399-acf2-340c8e3e6c05-0"	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: running
[...] (workflowStep)	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello Node: Component	 Operation: custom.say_hello	 Status: done
[...] (Execution)	 Execution: "0d00bc24-fb5b-4399-acf2-340c8e3e6c05-0"	 Workflow: wf2	 Instance: 0	 Step: Component_say_hello	 Node: Component	 Operation: custom.say_hello	 Status: done
[...] (workflow)	 Workflow: wf2	 Status: done

```
Check workflow wf2 task info (task id provided in the first event of the workflow execution above) :

```
yorc deployments tasks info --steps test 6d1bd3b6-9bd0-4c4f-8ac4-bf91eb87ea3e

Task:  6d1bd3b6-9bd0-4c4f-8ac4-bf91eb87ea3e
Task status: DONE
Task type: CustomWorkflow
Steps:
+---------------------+--------+
| Name                | Status |
+---------------------+--------+
| Component_say_hello | DONE   |
| inline_wf1          | DONE   |
+---------------------+--------+
```

### Description for the changelog

Failure running workflow with inline activity ([GH-592](https://github.com/ystia/yorc/issues/592))

## Applicable Issues

Closes #592

### CODE REVIEW

File: tasks/workflow/step.go [hunk 1]
```diff
@@ -371,22 +372,48 @@ func (s *step) runActivity(wfCtx context.Context, cfg config.Configuration, depl
...
+	data[taskDataParentTaskID] = s.t.taskID
+	data[taskDataDeploymentID] = deploymentID
+	data[taskDataWorkflowName] = workflowName
+
+	taskID, err := collector.NewCollector(s.cc).RegisterTaskWithData(deploymentID, tasks.TaskTypeCustomWorkflow, data)
 	if err != nil {
-		err = errors.Wrapf(err, "Failed to register workflow init operations with workflow:%q, targetID:%q, taskID:%q", workflowName, s.t.targetID, s.t.taskID)
+		err = errors.Wrapf(err, "Failed to register inline workflow %s in parent workflow %s step %s, targetID %s, taskID %s",
+			workflowName, s.WorkflowName, s.Name, s.t.targetID, s.t.taskID)
+	} else {
```
Comment (loicalbertin) [inline tasks/workflow/step.go]: this else block is not very straightforward. 
If the idea is just to save a `if err != nil` I think it doesn't worthwhile the lake of readability.

File: tasks/workflow/workflow.go [hunk 2]
```diff
@@ -107,3 +121,107 @@ func updateTaskStatusAccordingToWorkflowStatus(ctx context.Context, deploymentID
 	}
 	return status, errors.Wrapf(checkAndSetTaskStatus(ctx, deploymentID, taskID, status), "Failed to update task status to %q with TaskID: %q", status, taskID)
 }
+
+// Get the parent workflow of an inline workflow, returns an empty string if there
+// is no parent workflow
+func getParentWorfklow(ctx context.Context, t *taskExecution, wfName string) (string, error) {
```
Comment (stebenoist) [inline tasks/workflow/workflow.go]: typo here "getParentWorfklow"

File: tasks/workflow/step.go [hunk 3]
```diff
@@ -371,22 +372,48 @@ func (s *step) runActivity(wfCtx context.Context, cfg config.Configuration, depl
...
+		err = s.setStatus(tasks.TaskStepStatusRUNNING)
 	}
 
+	if err != nil {
+		err = s.setStatus(tasks.TaskStepStatusERROR)
+
+	}
+	// Marking this step as asynchronous as it should not be considered as
+	// done by the caller
+	s.Async = true
```
Comment (stebenoist) [inline tasks/workflow/step.go]: Does async step needs any monitoring action ?

Comment (laurentganne) [reply tasks/workflow/step.go]: No, the inline workflow when it ends will take care of updating this parent step in parent workflow

File: tasks/workflow/step.go (hunk 1 above)
Comment (laurentganne) [reply tasks/workflow/step.go]: yes. Ok removing

File: tasks/workflow/workflow.go (hunk 2 above)
Comment (laurentganne) [reply tasks/workflow/workflow.go]: good catch

File: tasks/workflow/step.go (hunk 3 above)
Comment (stebenoist) [reply tasks/workflow/step.go]: OK

**Pull request merged**